src/log.c

```diff
@@ -0,0 +1,361 @@
+    y001 = (ad - y400*YEAR400 - y100*YEAR100 - y004*YEAR004)/YEAR001;
+    yy = y400*4*100 + y100*100 + y004*4 + y001*1 + YEARFIRST;
+    dd = (ad - y400*YEAR400 - y100*YEAR100 - y004*YEAR004)%YEAR001;
+
+    if(0 == yy%1000)
+    {
+        if(0 == (yy/1000)%4)
+        {
+            m[1] = 29;
+        }
+    } else {
+        if(0 == yy%4)
+        { 
+            m[1] = 29;
+        }
+    }
+    for(i = 1; i <= 12; i++)
+    {
+        if(dd - m[i] < 0)
+        {
+            break;
+        } else {
+            dd = dd -m[i];
+        }
+    }
+    
+    mm = i;
+    hh = sec/(60*60)%24;
+    mi = sec/60 - sec/(60*60)*60;
+    ss = sec - sec/60*60;
+    snprintf(time, time_len, "%d-%02d-%02d %02d:%02d:%02d\n", yy, mm, dd, hh, mi, ss);
+    return tm;
+}
+
+static int dpvs_async_log(uint32_t level, uint32_t logtype, lcoreid_t cid, char *log, int len, int off)
+{
+    struct dpvs_log *msg = NULL;
+    int log_hash_new;
+    int err;
+
+    log_hash_new = log_BKDRHash(log+off, len);
+    if (log_hash_new == log_stats_info[cid].log_hash 
+            && (rte_get_timer_cycles()-log_stats_info[cid].log_begin) < log_internal*rte_get_timer_hz()){  
+        log_stats_info[cid].missed++;    
+        return -1;
+    }
+
+    /* add time info and send out to log ring */ 
+    if (off) {
+        log_get_time(log, LOG_SYS_TIME_LEN);       
+        log[LOG_SYS_TIME_LEN-1] = ' '; 
+        len += LOG_SYS_TIME_LEN;
+    }
+    log_stats_info[cid].log_hash = log_hash_new;
+    log_stats_info[cid].log_begin = rte_get_timer_cycles();
+
+    msg = dpvs_log_msg_make(level, logtype, cid, len, log);
+    if (msg == NULL)
+        return -1;
+    
+    err = log_send(msg);
+    if (err != EDPVS_OK) {
+        dpvs_log_free(msg);
+        /* log ring is full, need to set limit rate */
+        fprintf(stderr, "log ring is full !\n");
+        log_stats_info[cid].slow = 1;
+        log_stats_info[cid].slow_begin = rte_get_timer_cycles();
+        return -1;
+    }
+    return 0;
+}
+
+int dpvs_log(uint32_t level, uint32_t logtype, const char *func, int line, const char *format, ...)
+{
+    va_list ap;
+    lcoreid_t cid;
+    char log_buf[DPVS_LOG_MAX_LINE_LEN];
+    int len = 0;
+    int off = g_dpvs_log_time_off;
+
+    if (level > rte_logs.level)
+        return -1;
+
+    va_start(ap, format);
+
+    do { 
+        if (!g_dpvs_log_async_mode || !g_dpvs_log_core || !g_dpvs_log_thread_ready) {
+            rte_vlog(level, logtype, format, ap);
+            break;
+        }
+
+        /* async log is not used for ctrl message */
+        if (logtype != RTE_LOGTYPE_USER1) {
+            rte_vlog(level, logtype, format, ap);
+            break;
+        }
+
+        cid = rte_lcore_id();
+
+        if (log_stats_info[cid].slow) {
+            /* set log limit rate to 5 sec and keep for 10 mins */
+            if (rte_get_timer_cycles() - log_stats_info[cid].slow_begin > LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()) {
```

Comment:
`LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()`
Keep the same spacing style for math operations for each side of comparsion.